dipy/align/mattes.pyx

```diff
@@ -0,0 +1,1276 @@
+#!python
+#cython: boundscheck=False
+#cython: wraparound=False
+#cython: cdivision=True
+
+import numpy as np
+cimport numpy as cnp
+cimport cython
+from fused_types cimport floating
+
+from dipy.align.vector_fields cimport(_apply_affine_3d_x0,
+                                      _apply_affine_3d_x1,
+                                      _apply_affine_3d_x2,
+                                      _apply_affine_2d_x0,
+                                      _apply_affine_2d_x1)
+
+from dipy.align.transforms cimport (Transform)
+
+cdef extern from "dpy_math.h" nogil:
+    double cos(double)
+    double sin(double)
+    double log(double)
+
+class MattesBase(object):
+    def __init__(self, nbins):
+        r""" MattesBase
+        Base class for the Mattes' Mutual Information metric.
+        This implementation is not tied to any optimization (registration)
+        method, the idea is that a registration metric based on MI should
+        inherit from this class to perform the low-level computations of
+        the joint intensity distributions and its gradient w.r.t. the
+        transform parameters, and then communicate the results to the
+        appropriate optimizer.
+
+        Notes: we need this class in cython to allow
+        _joint_pdf_gradient_dense_2d and _joint_pdf_gradient_dense_3d to
+        use a nogil Jacobian function (obtained from an instance of the
+        Transform class), which allows us to evaluate Jacobians at all
+        the sampling points (maybe the full grid) inside a nogil loop.
+
+        The reason we need a class is to encapsulate all the parameters
+        related to the joint and marginal distributions.
+
+        Parameters
+        ----------
+        nbins : int
+            the number of bins of the joint and margianl probability density
+            functions (the actual number of bins of the joint PDF is nbins^2)
```

Comment:
Fixed, used nbins**2. Sorry, this may be a silly question, but is the latex notation actually useful in the docstring, I mean, are the docstrings being parsed at some point to generate the math expressions?